### DISCUSSION TITLE
KV store

### DISCUSSION THREAD

**Discussion by nexovec** [maypok86/otter#107] [General]
Hi, is it possible to use otter as a KV store and if not, why not?

**Comment by maypok86** [comment]
Hi, otter is just an in-memory cache and stores everything in RAM. If you need a disk cache, then you can pay attention to [theine-go](https://github.com/Yiling-J/theine-go).

**Reply by nexovec** [reply]
If I don't care about disk persistence, is otter a good solution for K-V? I don't understand the implementation, so just wanted to verify.

**Reply by maypok86** [reply]
Most likely, yes. You just need to remember about the eviction of entries.

**Comment by nexovec** [comment]
here are some more questions that would help me decide:

- Is write order preserved? If not, is write order preserved on a single thread?
- Is it consistent? (If I write something, is it immediately available for reading?) If not, is it on a single thread?
- Is otter well suited for workloads that don't evict?
- Is it possible to do a disk snapshot and restore?

Also want to say that otter is a beautiful library and I love every bit of it(and have been working with it extensively), especially the API.

**Comment by maypok86** [comment]
> Is write order preserved? If not, is write order preserved on a single thread?

I didn't fully understand, but most likely not. You may think that otter is just a map with additional features.

> Is it consistent? (If I write something, is it immediately available for reading?) If not, is it on a single thread?

Yes, both on one thread and on many.

> Is otter well suited for workloads that don't evict?

Quite possibly, you can just specify capacity as unreachable, but most likely there are better solutions for this.

> Is it possible to do a disk snapshot and restore?

It's not out of the box, but you can implement it yourself based on `Range`.

**Reply by nexovec** [reply]
> you can implement it yourself based on Range.

I want to ask if it wouldn't be better to dump the actual byte buffer underneath otter into a file instead?

**Reply by maypok86** [reply]
> > you can implement it yourself based on Range.
> 
> I want to ask if it wouldn't be better to dump the actual byte buffer underneath otter into a file instead?

Perhaps yes, but the only good solution when saving the cache to a file is to save meta information, but saving meta information about the eviction policy to a file can lead to unpleasant errors when updating a library in which the structure of the eviction policy has been changed. If we do not save the eviction policy, then there is no great point in adding this feature to the library.

**Reply by nexovec** [reply]
@maypok86 there could just be a version tag, but then you'd still have to deal with migrations, that's probably not worth it right?

**Reply by maypok86** [reply]
Yes, it all looks like overengineering. The elements that we want to cache rarely change in a short period of time, so for now I try to always vote for developers to simply save the cache to a file via Range. If they need it, of course.

**Comment by nexovec** [comment]
I'm happy the answer seems to be "yes, of course."

Are you interested in merging the snapshot feature?
